### PR TITLE
Improve context for Known false positives

### PR DESCRIPTION
**Pull request by YosiElias**
Organized the VDB so that each vector contains the full known issue, with metadata including the reason for being classified as a FP and the issue type.

**Example metadata:**
`metadata = {'reason_of_false_positive': '# there is a check for k<0', 'issue_type': 'INTEGER_OVERFLOW'}`

The retriever first filters by issue type and then returns the N most similar issues (where N is defined in the environment variable NUMBER_OF_RETRIEVED_KNOWN_ISSUE).

**Results:**

The retriever gave **a perfect match** on the known issue context side, here is the [results](https://docs.google.com/spreadsheets/d/1O8TLqfjfGN9F8zCYr33GOUsVy3gppRm4/edit?usp=sharing&ouid=111619790710498255381&rtpof=true&sd=true).

**Edge Cases Tested:**

- def0: A fake issue created to test the longest known issue in a single vector.
- def7: Very similar to a known issue but not an exact match.
- def25: Contains an issue type that does not exist in the known issues.
- The remaining issues are all known issues.

### CODE REVIEW

File: src/LLMService.py [hunk 1]
```diff
@@ -87,39 +92,50 @@ def critique_llm(self):
...
-             "answer template."
-             "\nAnswer response template:{answer_template}"
-             "\n\nContext:{context}"
-             ),
+            "You are an expert in identifying similar error stack traces. "
+            "The context contains a list of issues that have been classified as false positive by RedHat engineers. "
+            "Each issue in the list includes two key elements: "
+            "1. An error trace (called 'Known False Positive'). "
+            "2. A reason for its classification as a false positive (called 'Reason Marked as False Positive'). "
+            "When comparing issues, you may ignore differences in line numbers and systemd version. "
```
Comment (JudeNiroshan) [inline src/LLMService.py:104]: Can we not mention about the word "systemd". Because when we run the pipeline for another project, we also need to change this line. Think about a different approach we can take.

File: src/LLMService.py [hunk 2]
```diff
@@ -87,39 +92,50 @@ def critique_llm(self):
...
+            "Answer the question using only the provided context. "
+            "Your response must strictly follow the provided answer response template. "
+            "Do not include any additional text outside the answer template. "
+            "\nAnswer response template:{answer_template}"
+            "\n\nContext:{context}"
+            ),
             ("user", "{question}")
         ])
-        retriever = database.as_retriever()
+        retriever = database.as_retriever(search_kwargs={"k": NUMBER_OF_RETRIEVED_KNOWN_ISSUE, 'filter': {'issue_type': issue.issue_name}})
```
Comment (JudeNiroshan) [inline src/LLMService.py:117]: can you make it multiline obj representation. It would be easy to read.

File: src/LLMService.py [hunk 3]
```diff
@@ -1,4 +1,6 @@
 import os
+from pprint import pprint
```
Comment (JudeNiroshan) [inline src/LLMService.py:2]: Can you also adjust other places where we simply used `print("".center(80, '-'))`?

File: src/LLMService.py [hunk 4]
```diff
@@ -87,39 +92,50 @@ def critique_llm(self):
...
+            ),
             ("user", "{question}")
         ])
-        retriever = database.as_retriever()
+        retriever = database.as_retriever(search_kwargs={"k": NUMBER_OF_RETRIEVED_KNOWN_ISSUE, 'filter': {'issue_type': issue.issue_name}})
         resp = retriever.invoke(user_input)
-        context_str = "".join(doc.page_content for doc in resp)
+        context_list = self._format_context_from_response(resp)
+        print(f"[issue-ID - {issue.id}] Found This context:")
+        pprint(context_list, width=100)
```
Comment (JudeNiroshan) [inline src/LLMService.py:121]: Is it really necessary to use this external dependency? If the answer is yes, then fine.

Comment (JudeNiroshan) [reply src/LLMService.py:121]: Also, for every issue, we are printing the context_list. I think this will add unnecessary noise in the logs.

File: src/LLMService.py [hunk 5]
```diff
@@ -87,39 +92,50 @@ def critique_llm(self):
...
-                    "context": RunnableLambda(lambda _: context_str),
+                    "context": RunnableLambda(lambda _: context_list),
                     "answer_template": RunnableLambda(lambda _: answer_template),
                     "question": RunnablePassthrough()
                 }
                 | prompt
         )
         actual_prompt = chain1.invoke(user_input)
+        if not context_list:
+            print(f"Not find any relevant context for issue id {issue.id}")
```
Comment (JudeNiroshan) [inline src/LLMService.py:136]: Since we are printing when the issue is found in the known false positive case, we do not need this print statement.

File: src/LLMService.py [hunk 6]
```diff
@@ -225,11 +249,38 @@ def _evaluate(self, actual_prompt, response):
         return critque_response
 
     def create_vdb(self, text_data):
-        self.embedding_llm.embed_documents(text_data)
         self.vector_db = FAISS.from_texts(text_data, self.embedding_llm)
-        # self.vector_db.add_texts(text_data)
         return self.vector_db
 
+    def create_vdb_for_knonw_issues(self, text_data):
```
Comment (JudeNiroshan) [inline src/LLMService.py:255]:
```suggestion
    def create_vdb_for_known_issues(self, text_data):
```

File: src/LLMService.py [hunk 7]
```diff
@@ -225,11 +249,38 @@ def _evaluate(self, actual_prompt, response):
...
+        metadata_list, error_trace_list = self._process_knonw_issues(text_data)
```
Comment (JudeNiroshan) [inline src/LLMService.py:256]:
```suggestion
        metadata_list, error_trace_list = self._process_known_issues(text_data)
```

File: src/templates/unknown_issue_filter_resp.json [hunk 8]
```diff
@@ -0,0 +1,8 @@
+{
+  "Issue_id": "{ID}",
```
Comment (JudeNiroshan) [inline src/templates/unknown_issue_filter_resp.json:2]: Shall stick to the JSON attribute naming convention of `snake_case`? Just to be consistence across the project.

File: src/run.py [hunk 9]
```diff
@@ -126,20 +126,29 @@
...
-
-        # let's calculate numbers for quality of the response we received here!
+        
         score = {}
-        if CALCULATE_METRICS:
-            metric_request = metric_request_from_prompt(prompt, response)
-            score = metric_handler.evaluate_datasets(metric_request)
+        if issue.id in already_seen_issue_ids.keys():
+            print(f"{issue.id} already marked as a false positive since it's a known issue")
+            template_path = os.path.join(os.path.dirname(__file__), "templates", "final_jude_resp.json")
```
Comment (JudeNiroshan) [inline src/run.py:133]:
```suggestion
            template_path = os.path.join(os.path.dirname(__file__), "templates", "final_resp.json")
```

File: src/LLMService.py (hunk 2 above)
Comment (Yael-F) [inline src/LLMService.py:117]: Really like the filtering addition — great improvement!

To make things a bit clearer, could you update the Issue class to store the issue type as 'issue_type' instead of 'issue_name'? That way, it’ll be easier to follow throughout the code.

File: src/LLMService.py [hunk 10]
```diff
@@ -87,39 +92,50 @@ def critique_llm(self):
...
                 {
-                    "context": RunnableLambda(lambda _: context_str),
+                    "context": RunnableLambda(lambda _: context_list),
                     "answer_template": RunnableLambda(lambda _: answer_template),
                     "question": RunnablePassthrough()
                 }
                 | prompt
         )
         actual_prompt = chain1.invoke(user_input)
+        if not context_list:
```
Comment (Yael-F) [inline src/LLMService.py:135]: Would it make sense to move this check to be right after the retriever? 
I guess we don't really need the prompt in this case - and placing it earlier might make the flow a bit cleaner :)

File: src/Utils/file_utils.py [hunk 11]
```diff
@@ -15,10 +16,8 @@ def read_source_code_file(path):
 
 def read_known_errors_file(path):
     with open(path, "r", encoding='utf-8') as f:
-        text_splitter = RecursiveCharacterTextSplitter(separators=["\n\n", "\n"],
-                                                       chunk_size=500, chunk_overlap=0)
         plain_text = f.read()
-        doc_list = text_splitter.create_documents([plain_text])
+        doc_list = [item.strip() for item in plain_text.split("\n\n\n") if item.strip()!='']
```
Comment (Yael-F) [inline src/Utils/file_utils.py:20]: Since we are not limiting the size of the doc here, 
could you add a check to ensure the total doc size doesn’t exceed the embedding model's token limit?

**Review by Yael-F — CHANGES_REQUESTED** [review]
Great change!
left a few small comments, Let me know what you think :)

File: src/LLMService.py (hunk 1 above)
Comment (YosiElias) [reply src/LLMService.py]: [Done](https://github.com/RHEcosystemAppEng/sast-ai-workflow/pull/13/commits/3262c4b9a4a348b11392d4732bc0c185412fc9c2)

File: src/LLMService.py (hunk 2 above)
Comment (YosiElias) [reply src/LLMService.py]: [Done](https://github.com/RHEcosystemAppEng/sast-ai-workflow/pull/13/commits/3262c4b9a4a348b11392d4732bc0c185412fc9c2)

File: src/LLMService.py (hunk 4 above)
Comment (YosiElias) [reply src/LLMService.py]: I removed this dependency. I added this print because we don’t have an option to see this context like we can with the main context (SHOW_FINAL_JUDGE_CONTEXT). 

I’ll disable it but keep it for debugging. (BTW, I think we should add a logging mechanism with log levels to avoid all these disabled prints. WDYT?)

File: src/LLMService.py (hunk 5 above)
Comment (YosiElias) [reply src/LLMService.py]: It’s showing that the context is **empty** - not just that it didn’t find a similar known issue, but that there’s no issue of this type in the context at all. Again, I think this can be useful for debugging since the filter’s context is completely behind the scenes.
I'll disable this line too and keep for debug.

File: src/templates/unknown_issue_filter_resp.json (hunk 8 above)
Comment (YosiElias) [reply src/templates/unknown_issue_filter_resp.json:2]: Not sure I follow, how do you want to name it?

File: src/run.py (hunk 9 above)
Comment (YosiElias) [reply src/run.py]: [Done](https://github.com/RHEcosystemAppEng/sast-ai-workflow/pull/13/commits/3262c4b9a4a348b11392d4732bc0c185412fc9c2)

File: src/LLMService.py (hunk 2 above)
Comment (YosiElias) [reply src/LLMService.py]: The issue_type field on Issue object only contains the word "Error," so I [changed](https://github.com/RHEcosystemAppEng/sast-ai-workflow/pull/13/commits/3262c4b9a4a348b11392d4732bc0c185412fc9c2) it to hold the type instead of issue_name.
Is it ok from your side, @JudeNiroshan ?

File: src/LLMService.py (hunk 4 above)
Comment (JudeNiroshan) [reply src/LLMService.py]: Yes, let's take it as a different task

File: src/templates/unknown_issue_filter_resp.json (hunk 8 above)
Comment (JudeNiroshan) [reply src/templates/unknown_issue_filter_resp.json:2]: Attribute names should be snake_case. e.g.: "Issue_id" -> "issue_id"

File: src/LLMService.py (hunk 2 above)
Comment (JudeNiroshan) [reply src/LLMService.py]: I'm fine. Change the code for more clarity.

File: src/Utils/file_utils.py (hunk 11 above)
Comment (YosiElias) [reply src/Utils/file_utils.py:20]: For now, I have [added](https://github.com/RHEcosystemAppEng/sast-ai-workflow/pull/13/commits/a7e98f26b96d46f215a4facf068b2f24db6d5f3d) a warning. Based on the edge cases I reviewed in our data, we do not currently reach the size limit. However, for future safety, I agree that we should monitor this. Thank you for the reference code and examples, @Yael-F !

File: src/templates/unknown_issue_filter_resp.json (hunk 8 above)
Comment (YosiElias) [reply src/templates/unknown_issue_filter_resp.json]: [Done](https://github.com/RHEcosystemAppEng/sast-ai-workflow/pull/13/commits/23b01bdc443e5d2dcc2e98b6a247b6630c202363), thanks for your comments @JudeNiroshan !
